sharing/loop_through_rules.py

```diff
@@ -0,0 +1,229 @@
+    defines a bracket type. Determines the limit (lower and upper) for range
+    type or the single filter value.
+    If the rule is of datetime type, then loops through the datetime variables.
+    If the rule value is of numeric type, then loops through the numeric
+    variables of current rule.
+
+    The function returns the the filtered data for current table.
+
+    Parameters:
+        rule(dict): current organization rule
+        list_of_rules(list): rule in current ruleValue sharing property
+        intermediate_filtered_data(dataframe): data to be filtered for each table
+        datetime_variables(list): current rule variables list of type datetime
+        numeric_variables(list): current rule variables list of type numeric
+        string_variables(list): current rule variables list of type string
+
+    Returns:
+        dataframe: filtered data for current rule table
+
+    """
+
+    # Setting patterns for detecting type of brackets
+    pat3: str = r"^\["
+    pat4: str = r"\]$"
+    pat5: str = r"^\("
+    pat6: str = r"\)$"
+
+    # ITTERATING THROUGH EACH RULEVALUE
+    for filter_rule in list_of_rules:
+
+        filter_by: Union[str, int, float] = filter_rule
+        if isinstance(filter_rule, str):
+            filter_by = filter_rule.strip()
+
+        # Assign a variable for the bracket type, defines lower and upper limit
+        # if the particular ruleValue is a range type
+        bracket: str = ""
+        lower_limit: str = ""
+        upper_limit: str = ""
+
+        # Assign a variable for the ruleValue that is not a range type
+        filter_val: Union[str, float, int, None] = None
+
+        if isinstance(filter_by, str):
+
+            # Determine the type of bracket present in the ruleValue if it is a range type.
+            if (
+                re.search(pat3, filter_by) is not None
+                and re.search(pat4, filter_by) is not None
+            ):
+                bracket = "[]"
+                print("BRACKET PRESENT []")
+            if (
+                re.search(pat5, filter_by) is not None
+                and re.search(pat6, filter_by) is not None
+            ):
+                bracket = "()"
+                print("BRACKET PRESENT ()")
+            if (
+                re.search(pat5, filter_by) is not None
+                and re.search(pat4, filter_by) is not None
+            ):
+                bracket = "(]"
+                print("BRACKET PRESENT (]")
+            if (
+                re.search(pat3, filter_by) is not None
+                and re.search(pat6, filter_by) is not None
+            ):
+                bracket = "[)"
+                print("BRACKET PRESENT [)")
+
+        # Confirm if the ruleValue is range type by checking if bracket exist
+        # & create upper_limit and lower_limit value
+        if bracket and isinstance(filter_by, str):
+            if "," in filter_by:
+                filter_list = filter_by.split(",")
```

Comment:
Should start by replacing the brackets by empty strings using the regexes defined at the beginning. This way you do not have to replicate the code in lines 107 and 111.